#6155: OAuth 2.0 - Microsoft

alejandratenorio opened the pull request:
## New Contributors

Welcome! New contributors should at least glance at [CONTRIBUTING.md](/CONTRIBUTING.md), especially the section on pull requests where we encourage you to reach out to other developers before you start coding. Also, please note that we measure code coverage and prefer you write unit tests. Pull requests can still be reviewed without tests or completion of the checklist outlined below. Thanks!

## Related Issues

- Closes #6155 

## Pull Request Checklist

- [ ] Unit [tests][] completed
- [ ] Integration [tests][]: None
- [ ] Deployment requirements, [SQL updates][], [Solr updates][], etc.: None
- [ ] [Documentation][docs] completed
- [ ] Merged latest from "develop" [branch][] and resolved conflicts

[tests]: http://guides.dataverse.org/en/latest/developers/testing.html
[SQL updates]: http://guides.dataverse.org/en/latest/developers/sql-upgrade-scripts.html
[Solr updates]: https://github.com/IQSS/dataverse/blob/develop/conf/solr/7.3.0/schema.xml
[docs]: http://guides.dataverse.org/en/latest/developers/documentation.html
[branch]: http://guides.dataverse.org/en/latest/developers/branching-strategy.html

djbrooke commented:
Thanks @alejandratenorio for the PR. We'll review this and I'll also add some documentation.

poikilotherm commented:
Since the merge of IQSS/dataverse-kubernetes#87, I am working on #5991 and it's recent PR about updating ScribeJava. It might be a good idea to have that merged first, so it doesn't have to be me testing the Microsoft OAuth2... :wink:

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/AbstractOAuth2AuthenticationProvider.java:
```diff
@@ -116,6 +116,10 @@ public OAuth2UserRecord getUserRecord(String code, String state, String redirect
         request.setCharset("UTF-8");
+
+        // Microsoft
+        request.addHeader("Accept", "application/json");
```
I'm slightly concerned about this because doesn't ORCID use XML instead of JSON?

Gerafp replied:
13:48
Hi!

We use this line because Microsoft identity platform needs this header when send 
an authentication request. If this header is out, we receive an error.

![image](https://user-images.githubusercontent.com/25359497/65272398-980e8d80-dae4-11e9-9b1c-c8b26660a6ce.png)

We work in a fix for the ORCID XML problem.

pdurbin replied:
@Gerafp perfect. Thanks! Please let us know when that fix is in.

djbrooke replied:
Thanks @Gerafp - I'll move this PR back to Community Dev.

Gerafp replied:
Hi.

We add a instruction for manage the authentication request when use Microsoft.

The branch has updated with our fix.

:D

djbrooke commented:
Hi @Gerafp and @alejandratenorio, can one of you please give me permission to push to this branch? I have some documentation edits ready to push. Thanks!

alejandratenorio commented:
> Hi @Gerafp and @alejandratenorio, can one of you please give me permission to push to this branch? I have some documentation edits ready to push. Thanks!

@djbrooke,  sure, it's done.

pdurbin commented on doc/sphinx-guides/source/installation/oauth2.rst:
```diff
@@ -11,7 +11,7 @@ As explained under "Auth Modes" in the :doc:`config` section, OAuth2 is one of t
 
 `OAuth2 <https://oauth.net/2/>`_ is an authentication protocol that allows systems to share user data, while letting the users control what data is being shared. When you see buttons stating "login with Google" or "login through Facebook", OAuth2 is probably involved. For the purposes of this section, we will shorten "OAuth2" to just "OAuth." OAuth can be compared and contrasted with :doc:`shibbo ...
 
-Dataverse supports three OAuth providers: `ORCID <http://orcid.org>`_, `GitHub <https://github.com>`_, and `Google <https://console.developers.google.com>`_.
```
@djbrooke should we add "Microsoft" to the title of this page? Right now line 1 is "OAuth Login: ORCID, GitHub, Google".

djbrooke replied:
Updated in https://github.com/IQSS/dataverse/pull/6192/commits/1bf38c2ee36dec99e667289198ea768228519924

djbrooke commented:
Thanks @Gerafp and @alejandratenorio for giving me push access! Please take a look at the docs I added and feel free to expand those sections if necessary. Also note that I did not add any info about the json template at line 55 in oauth.rst, as it seems that still needs to be finalized. Thanks again for this contribution!

pdurbin commented:
> Please be aware that in current releases of ScribeJava an upstream API client is available for this Microsoft endpoint.

I'm a little confused.

Currently Dataverse uses ScribeJava 3.1.0. According to https://github.com/scribejava/scribejava/tree/scribejava-3.1.0#supports-all-major-10a-and-20-oauth-apis-out-of-the-box the following Microsoft API is supported in 3.1.0:

- Microsoft Live (https://login.live.com)

If we upgrade to ScribeJava 6.8.1 (the latest), https://github.com/scribejava/scribejava/tree/scribejava-6.8.1#supports-all-major-10a-and-20-oauth-apis-out-of-the-box indicates that a total of three Microsoft APIs are supported:

- Microsoft Azure Active Directory (Azure AD) (http://azure.microsoft.com/)
- Microsoft Azure Active Directory (Azure AD) 2.0 (http://azure.microsoft.com/)
- Microsoft Live (https://login.live.com/)

Which of these three Microsoft APIs will Dataverse support when this pull request is merged? All of them? One of them?

alejandratenorio commented:
> Thanks @Gerafp and @alejandratenorio for giving me push access! Please take a look at the docs I added and feel free to expand those sections if necessary. Also note that I did not add any info about the json template at line 55 in oauth.rst, as it seems that still needs to be finalized. Thanks again for this contribution!

@djbrooke, I added info about microsoft.json file.

djbrooke commented:
I took this issue because I have some questions about the plans to sustain this code post-merge and there are some questions about the best way to test this. Will discuss post standup in 30 minutes. 

@alejandratenorio, did you use https://oauthplay.azurewebsites.net/ in your testing or some other test system?

djbrooke commented:
@alejandratenorio - can you also respond to @pdurbin's questions above in https://github.com/IQSS/dataverse/pull/6192#issuecomment-533654103 ? This will help us in our testing, and it is important that we also understand as much as possible about the implementation. Thank you!

alejandratenorio commented:
> If we upgrade to ScribeJava 6.8.1 (the latest),

Hi @pdurbin,

I'm sorry I'm late, With ScribeJava 3.1.0 only supports https://login.live.com/.

alejandratenorio commented:
> I took this issue because I have some questions about the plans to sustain this code post-merge and there are some questions about the best way to test this. Will discuss post standup in 30 minutes.
> 
> @alejandratenorio, did you use https://oauthplay.azurewebsites.net/ in your testing or some other test system?



Hi, @djbrooke 

For testing we use https://apps.dev.microsoft.com/#/appList
now redirects to
https://login.microsoftonline.com/common/oauth2/authorize

djbrooke commented:
Hi @alejandratenorio, @Gerafp,

Thanks for continuing to communicate about this. It would be good to have a meeting about this to discuss. The goals of this meeting would be to:

- Get on the same page about what's being implemented and the method of implementation. There are a lot of options here and I want to make sure we're understanding and testing appropriately. This should include a demo of what you have done.
- How to best document this
- Discussions about how to best maintain this code going forward
- ScribeJava upgrade implications 

Are you available at 2 PM ET on Thursday or Friday of this week? I'll set up a remote meeting once you confirm. 

Thanks again!

poikilotherm commented:
I would be very happy to join the discussion and share my activities of almost finished and 2/3 tested #5997, but timezones are killing me.

djbrooke commented:
I hear you @poikilotherm, thanks for the reminder. :) 

@alejandratenorio, @Gerafp, would 10 AM ET on either Monday the 30th or Tuesday the 1st work for you instead of the above mentioned times?

alejandratenorio commented:
> I hear you @poikilotherm, thanks for the reminder. :)
> 
> @alejandratenorio, @Gerafp, would 10 AM ET on either Monday the 30th or Tuesday the 1st work for you instead of the above mentioned times?

@djbrooke, It's okay on Monday the 30th at 10:00. Will we use Skype, Zoom, etc?

djbrooke commented:
Let's use this GoToMeeting link to meet at 10 AM on the 30th:

https://global.gotomeeting.com/join/911411085

alejandratenorio commented:
> Let's use this GoToMeeting link to meet at 10 AM on the 30th:
> 
> https://global.gotomeeting.com/join/911411085

@djbrooke, Gracias.

poikilotherm commented:
As I kind of self-invited me: are you guys ok with me joining the chat on Monday? Have a nice weekend! :smiley:

djbrooke commented:
Yes @poikilotherm, it's fine! Have a good weekend!

alejandratenorio commented:
Hi @djbrooke,

To limit access to personal Microsoft accounts or work/school accounts, we could add more attributes in the JSON configuration file, for example,
	-> Azure AD tenant, 'common' by default, https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-v2-protocols#endpoints
	-> domain, to filter domains email accounts. 

What do you think?

djbrooke commented:
Hi @alejandratenorio and @Gerafp - we discussed this in a design meeting this morning and I think that instead of making this part of a larger redesign project for the login options, we'll be able to move more quickly than we thought. The general steps here are:

- @scolapasta and @poikilotherm will discuss #5997 to determine whether or not it should get merged in before this or after this
- We need to re-evaluate what we're titling the different login options. I'll work with @TaniaSchlatter and whoever else is interested about what to title the various login options (based on the work started in the spreadsheet here https://docs.google.com/spreadsheets/d/10BvCESZor5iXqUV9EGJUAPtCkaD6pEXx-A7x61yeqak/edit#gid=0). 
- We'll take this through code review and QA and get it merged

djbrooke commented:
Hi @alejandratenorio, @Gerafp, thanks for your patience on this. @scolapasta and @poikilotherm should be able to give you some more technical guidance next week, but I just talked with Tania about this and we have some guidance. 

- We'd expect this to look similar to how Google and Github are implemented on slides 8 and 9 here. The user would select the "Microsoft" button on the bottom of the page and then be presented with a button that would take them into the Microsoft workflow. https://docs.google.com/presentation/d/1DmRPaLf19rdyDugWHuwiUoC7RoZDLlbogIqpfn2xdUU/edit#slide=id.g61be9907be_1_15
- This may require us to change how some things are labeled and named in the application. If you put up the PR (following the guidance from @scolapasta), we'll be able to make the changes to the labels on the buttons. 

If you have any questions, please feel free to let me know. Thanks again for this contribution!

Gerafp commented:
> Hi @alejandratenorio, @Gerafp, thanks for your patience on this. @scolapasta and @poikilotherm should be able to give you some more technical guidance next week, but I just talked with Tania about this and we have some guidance.
> 
> * We'd expect this to look similar to how Google and Github are implemented on slides 8 and 9 here. The user would select the "Microsoft" button on the bottom of the page and then be presented with a button that would take them into the Microsoft workflow. https://docs.google.com/presentation/d/1DmRPaLf19rdyDugWHuwiUoC7RoZDLlbogIqpfn2xdUU/edit#slide=id.g61be9907be_1_15
> * This may require us to change how some things are labeled and named in the application. If you put up the PR (following the guidance from @scolapasta), we'll be able to make the changes to the labels on the buttons.
> 
> If you have any questions, please feel free to let me know. Thanks again for this contribution!

Hi! @djbrooke 

We check our implementation and we see it looks like that

![image](https://user-images.githubusercontent.com/25359497/66686994-d3652d80-ec46-11e9-8777-74ccaa854bb4.png)

Are you referring to this when say buttons and labels? :D

djbrooke commented:
Thanks @Gerafp for confirming! That's great, that's what we would expect. I will make any needed changes to the labels and buttons.

djbrooke commented:
Thanks @poikilotherm. Do you still need sandbox access from ORCiD?

djbrooke commented:
Thanks @poikilotherm we'll see what we can do.

scolapasta commented:
@poikilotherm and I had discussed his status separately and I see he added his status in a comment above. Great! Based on the fact that PR #5997 is "95% done", we should get that work in first. One option is to wait until that gets merged and then grab the latest from develop, the other is to start getting those changes now (while waiting for the Orcid part to be tested). @poikilotherm should be able to help with what is needed to do this.

djbrooke commented:
@alejandratenorio I'm going to move this back to Community Dev for now. Do you plan to wait until #5997 is merged to develop or do you plan to use the code that's in progress?

alejandratenorio commented:
> Do you plan to wait until #5997 is merged to develop or do you plan to use the code that's in progress?

@djbrooke, I think it's better to wait until #5997 is merged, only to plan our activities, do you have an estimated end date?

djbrooke commented:
@alejandratenorio, it's in QA right now, so hopefully it will be merged soon. I'd suggesting checking in on https://github.com/orgs/IQSS/projects/2 early next week. I'll also do my best to notify here when it's complete. Thank you for being flexible with this!

djbrooke commented:
Hey @alejandratenorio and @Gerafp! #5997 was just merged. Feel free to get the code from develop and please let us know when you'd like us to take another look at this.

alejandratenorio commented:
> Hey @alejandratenorio and @Gerafp! #5997 was just merged. Feel free to get the code from develop and please let us know when you'd like us to take another look at this.

@djbrooke, thank you, we'll let you know.

pdurbin commented:
@alejandratenorio and @Gerafp now that pull request #6294 has been merged, can you please merge the latest from the develop? If you already merged the latest from develop I'm sorry for any disruption this causes. If you have trouble deploying, you can drop your database and start over or delete the old migration from your `flyway_schema_history` database table. Please feel free to ping me at http://chat.dataverse.org if you have any issues.

alejandratenorio commented:
> can you please merge the latest from the develop

Hi @pdurbin, we merged with the develop branch but we have some conflicts, we are fixing them.

Gerafp commented:
Hi all!

We update the MicrosoftOauth2AP class for ScribeJava 6.6.3.

djbrooke commented:
Thanks @Gerafp and @alejandratenorio. Is it ready for us to Code Review again? 

Can you please also remind me what sandbox(es) and/or process you have used for testing?

poikilotherm commented:
@djbrooke feel free to re-request a review from me once this is in Code Review. I'm kinda "in" that code now... :wink:

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/AbstractOAuth2AuthenticationProvider.java:
```diff
@@ -127,20 +127,36 @@ public OAuth2UserRecord getUserRecord(String code, @NotNull OAuth20Service servi
         service.signRequest(accessToken, request);
     
         Response response = service.execute(request);
+//>>>>>>> develop
```
Should we remove all these `<<<<<<< HEAD` (and related) comments? These are from resolving merge conflicts?

Gerafp replied:
The branch has been update.

Gerafp commented:
> Thanks @Gerafp and @alejandratenorio. Is it ready for us to Code Review again?
> 
> Can you please also remind me what sandbox(es) and/or process you have used for testing?

Yes, We are ready for Code Review.

The sandbox is here

https://datatest.servehttp.com
dataverseAdmin
qwerty12


We test with
* ORCID SandBox
* Google
For locate conflicts with others Login Options

For Microsoft login we test with
* Personal account (live.com.mx, outlook.com)
* Institucional account (cgiar.com, alumno.buap.mx)

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/AbstractOAuth2AuthenticationProvider.java:
```diff
@@ -127,19 +127,20 @@ public OAuth2UserRecord getUserRecord(String code, @NotNull OAuth20Service servi
         
         OAuth2AccessToken accessToken = service.getAccessToken(code);
-    
+        final String userEndpoint = getUserEndpoint(accessToken);
```
This userEndpoint variable doesn't seem to be used so I assume it can be deleted.

Gerafp replied:
Hi @pdurbin 

The microsoft.json don't have changes. The firts version of microsoft.json is used in this case too.

In case of userEndpoint variable is probably a old implementation of OAuth2, We don't know if other autenticathon service use this.

pdurbin replied:
@Gerafp ok, thanks, I still think QA might have questions but we'll see.

Here's a screenshot from my IDE (Netbeans) of the unused variable. It looks like you could use it below (line 138) if you want. Or you could delete it. Or you could do nothing. It's just something I noticed.

![Screen Shot 2019-10-23 at 5 08 58 PM](https://user-images.githubusercontent.com/21006/67434338-00ef9680-f5b8-11e9-9c9c-e61d1eb410b4.png)

pdurbin commented:
@Gerafp Is there a reason you deleted https://github.com/IQSS/dataverse/blob/develop/doc/Architecture/update-user-account-info.png ? It looks like the commit was 4828bf7

poikilotherm commented on src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/impl/MicrosoftOAuth2AP.java:
```diff
@@ -0,0 +1,60 @@
+        this.title = "Microsoft";
+        this.clientId = aClientId;
+        this.clientSecret = aClientSecret;
+        this.scope = Arrays.asList("openid User.Read");
```
After reading some docs for Microsoft Graph, I don't think this scope setting makes sense.

It should be sufficient to use `User.Read`. We don't use OIDC here at all and that `openid` scope does not include any attributes - it will only allow the use to login.

See docs:
- https://docs.microsoft.com/en-us/graph/permissions-reference#user-permissions
- https://docs.microsoft.com/en-us/graph/permissions-reference#openid-permissions

Gerafp replied:
Yes is sufficient to use User.Read, We delete openid option.

poikilotherm commented on src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/impl/MicrosoftOAuth2AP.java:
```diff
@@ -0,0 +1,60 @@
+                    response.getString("givenName", ""),
+                    response.getString("surname", ""),
+                    response.getString("userPrincipalName", ""),
+                    "", "");
```
This might be beyond scope, but it would be totally awesome trying to receive values for `affiliation` fields.

- https://docs.microsoft.com/en-us/graph/api/resources/organization?view=graph-rest-1.0
- https://docs.microsoft.com/en-us/graph/api/organization-get?view=graph-rest-1.0
- https://docs.microsoft.com/en-us/graph/permissions-reference#organization-permissions

Gerafp replied:
We view the options for receive these values and can implement this in the future. It's really interesting for us. :D

poikilotherm commented on src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/impl/MicrosoftOAuth2AP.java:
```diff
@@ -0,0 +1,60 @@
+                    response.getString("givenName", ""),
+                    response.getString("surname", ""),
+                    response.getString("userPrincipalName", ""),
+                    "", "");
```
Could the `position` field be filled with `jobTitle` claim already accessible with `User.Read`? Does that value make sense here?

poikilotherm commented on src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/impl/MicrosoftOAuth2AP.java:
```diff
@@ -0,0 +1,60 @@
+            String persistentUserId = response.getString("id");
+            String username = response.getString("userPrincipalName");
+            return new ParsedUserResponse(displayInfo, persistentUserId, username,
+                    (displayInfo.getEmailAddress().length() > 0 ? Collections.singletonList(displayInfo.getEmailAddress()) : Collections.emptyList() )
```
I know that the GitHub implementation is using this, too. I am really keen to see if this actually works, as the email response is retrieved as a string, not being converted to a list anywhere as far as I can see.

I don't know if this has ever been tested with the GitHub provider. It's a good idea to test this anyway, as I don't know what happens when you receive a JSON array of mails in `response.getString()`.

Gerafp replied:
Yes, We test this. :D

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/impl/MicrosoftOAuth2AP.java:
```diff
@@ -0,0 +1,60 @@
+            AuthenticatedUserDisplayInfo displayInfo = new AuthenticatedUserDisplayInfo(
+                    response.getString("givenName", ""),
+                    response.getString("surname", ""),
+                    response.getString("userPrincipalName", ""),
```
I didn't click any of the links above, but yes, if an email address is not provided, the user will be asked to fill it in before creating their Dataverse account.

Gerafp replied:
If you considered the `mail` claim is the best option for this problem, We can change it. 
An advantage for Microsoft Login is the possibility of get a user email directly from API compared with others services such as Orcid or GitHub.

Gerafp replied:
Hi!

We test with `mail` claim but is valid only for institutional accounts, unfortunaly the `jobTitle` claim is similar. 
If we considered all Microsoft accounts for login in Dataverse, We need use `userPrincipalName`.

pdurbin replied:
@Gerafp if you're happy with the code, I'm happy. I'm moving this to QA.

djbrooke commented on doc/sphinx-guides/source/installation/config.rst:
```diff
@@ -201,7 +201,7 @@ As for the "Remote only" authentication mode, it means that:
 - ``:AllowSignUp`` is set to "false" to prevent users from creating local accounts via the web interface.
 - ``:DefaultAuthProvider`` has been set to use the desired authentication provider
-- The "builtin" authentication provider has been disabled (:ref:`api-toggle-auth-provider`). Note that disabling the "builtin" authentication provider means that the API endpoint for converting an account from a remote auth provider will not work. Converting directly from one remote authentication provider to another (i.e. from GitHub to Google) is not supported. Conversion from remote is always  ...
+- The "builtin" authentication provider has been disabled (:ref:`api-toggle-auth-provider`). Note that disabling the "builtin" authentication provider means that the API endpoint for converting an account from a remote auth provider will not work. Converting directly from one remote authentication provider to another (i.e. from GitHub to Microsoft) is not supported. Conversion from remote is alwa ...
```
Thanks @poikilotherm. I wrote this and I'll change it.

djbrooke replied:
updated in https://github.com/IQSS/dataverse/pull/6192/commits/3ab5e259a6e097ca305e8b350a3be2dcc3015e38

djbrooke commented on doc/sphinx-guides/source/installation/oauth2.rst:
```diff
@@ -11,7 +11,7 @@ As explained under "Auth Modes" in the :doc:`config` section, OAuth2 is one of t
 `OAuth2 <https://oauth.net/2/>`_ is an authentication protocol that allows systems to share user data, while letting the users control what data is being shared. When you see buttons stating "login with Google" or "login through Facebook", OAuth2 is probably involved. For the purposes of this section, we will shorten "OAuth2" to just "OAuth." OAuth can be compared and contrasted with :doc:`shibbo ...
 
-Dataverse supports three OAuth providers: `ORCID <http://orcid.org>`_, `GitHub <https://github.com>`_, and `Google <https://console.developers.google.com>`_.
+Dataverse supports four OAuth providers: `ORCID <http://orcid.org>`_, `Microsoft <https://www.microsoft.com>`_, `GitHub <https://github.com>`_, and `Google <https://console.developers.google.com>`_.
```
Yes, I'll change it to this for now, unless someone has a better idea. I was punting on naming/labeling things until we determined exactly what we were implementing.

djbrooke replied:
Changed in https://github.com/IQSS/dataverse/pull/6192/commits/97a7549596ea170e59d242c9e49ed40701b4e042

djbrooke commented on doc/sphinx-guides/source/installation/oauth2.rst:
```diff
@@ -1,4 +1,4 @@
-OAuth Login: ORCID, GitHub, Google
+OAuth Login: ORCID, Microsoft, GitHub, Google
```
Changed in https://github.com/IQSS/dataverse/pull/6192/commits/97a7549596ea170e59d242c9e49ed40701b4e042

djbrooke commented on doc/sphinx-guides/source/installation/oauth2.rst:
```diff
@@ -24,18 +24,19 @@ Identity Provider Side
+Before OAuth providers will release information about their users (first name, last name, etc.) to your Dataverse installation, you must request a "Client ID" and "Client Secret" from them. In the case of GitHub and Google, this is as simple as clicking a few buttons and there is no cost associated with using their authentication service. ORCID and Microsoft, on the other hand, do not have an aut ...
 
-URLs to help you request a Client ID and Client Secret from the providers supported by Dataverse are provided below.  For all of these providers, it's a good idea to request the Client ID and Client secret using a generic account, perhaps the one that's associated with the ``:SystemEmail`` you've configured for Dataverse, rather than your own personal ORCID, GitHub, or Google account:
+URLs to help you request a Client ID and Client Secret from the providers supported by Dataverse are provided below.  For all of these providers, it's a good idea to request the Client ID and Client secret using a generic account, perhaps the one that's associated with the ``:SystemEmail`` you've configured for Dataverse, rather than your own personal Microsoft, ORCID, GitHub, or Google account:
```
updated in https://github.com/IQSS/dataverse/pull/6192/commits/a44bb2b9a7d5998ae092e754a35aed7ed3724c87

djbrooke commented on doc/sphinx-guides/source/user/account.rst:
```diff
@@ -22,12 +22,13 @@ Dataverse has been configured for one or more of the following log in options:
 - Username/Email and Password
 - Institutional Log In
 - ORCID
+- Microsoft
```
I'm OK being more precise here.

djbrooke replied:
updated in https://github.com/IQSS/dataverse/pull/6192/commits/31df5a2606029ed5be9144c9351a3b8849324628

djbrooke commented on doc/sphinx-guides/source/user/account.rst:
```diff
@@ -134,10 +135,10 @@ Convert your Dataverse account away from ORCID for log in
 
-GitHub and Google Log In
+Microsoft, GitHub, and Google Log In
 ~~~~~~~~~~~~~~~~~~~~~~~~~
```
updated in https://github.com/IQSS/dataverse/pull/6192/commits/31df5a2606029ed5be9144c9351a3b8849324628

djbrooke commented on doc/sphinx-guides/source/installation/oauth2.rst:
```diff
@@ -24,18 +24,19 @@ Identity Provider Side
 
 - Basic information about your Dataverse installation such as a name, description, URL, logo, privacy policy, etc.
-- OAuth2 Redirect URI (ORCID) or Authorization Callback URL (GitHub) or Authorized Redirect URIs (Google): This is the URL on the Dataverse side to which the user will be sent after successfully authenticating with the identity provider. This should be the advertised URL of your Dataverse installation (the protocol, fully qualified domain name, and optional port configured via the ``dataverse.sit ...
+- OAuth2 Redirect URI (ORCID) or Redirect URI (Microsoft) or Authorization Callback URL (GitHub) or Authorized Redirect URIs (Google): This is the URL on the Dataverse side to which the user will be sent after successfully authenticating with the identity provider. This should be the advertised URL of your Dataverse installation (the protocol, fully qualified domain name, and optional port config ...
```
updated in https://github.com/IQSS/dataverse/pull/6192/commits/31df5a2606029ed5be9144c9351a3b8849324628

djbrooke commented:
Thanks @poikilotherm for the thorough review here. I resolved all of the documentation issues and noted the commits. 

@Gerafp and @alejandratenorio, I'm going to move this back to community dev for you to respond to feedback and implement changes. If there's anything that you have questions about or anything that you feel is beyond the scope of your planned contribution, let's talk about it. Thanks again for working with us through the process. I think we're really close to getting this valuable feature tested, merged, and released.

poikilotherm commented on doc/sphinx-guides/source/user/account.rst:
```diff
@@ -134,10 +135,10 @@ Convert your Dataverse account away from ORCID for log in
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-You can also convert your Dataverse account to use authentication provided by GitHub or Google. These options may be found in the "Other options" section of the log in page, and function similarly to how ORCID is outlined above. If you would like to convert your account away from using one of these services for log in, then you can follow the same steps as listed above for converting away from th ...
+You can also convert your Dataverse account to use authentication provided by GitHub, Microsoft, or Google. These options may be found in the "Other options" section of the log in page, and function similarly to how ORCID is outlined above. If you would like to convert your account away from using one of these services for log in, then you can follow the same steps as listed above for converting  ...
```
@djbrooke now there is a mix of "Microsoft" and "Microsoft Azure AD". Is that helpfull for a user? Admins need precise technical things, but users will have "Microsoft" on that login button. Should we change that button, too? I'm not sure a user will know what "Microsoft Azure AD" is and what they can do with it.

djbrooke replied:
We will plan to rename the button, this is a big part of what I was punting. :) Not sure if Microsoft Azure AD button will make sense to users, either.

djbrooke replied:
Note to self: Maybe we'll just keep the button as Microsoft. Seems pretty common. 

<img width="534" alt="powerschool" src="https://user-images.githubusercontent.com/7773392/67514069-18875780-f66a-11e9-89f3-a2a6cec23f86.png">
<img width="485" alt="edx" src="https://user-images.githubusercontent.com/7773392/67514070-18875780-f66a-11e9-97e1-1054a2ac5bc2.png">
<img width="434" alt="wunderlist" src="https://user-images.githubusercontent.com/7773392/67514071-191fee00-f66a-11e9-965c-06eed005e9ae.png">
<img width="414" alt="doodle" src="https://user-images.githubusercontent.com/7773392/67514073-191fee00-f66a-11e9-935e-1ddbde1315ca.png">

poikilotherm commented on doc/sphinx-guides/source/installation/oauth2.rst:
```diff
@@ -1,4 +1,4 @@
-OAuth Login: ORCID, GitHub, Google
+OAuth Login: ORCID, Microsoft Azure Active Directory, GitHub, Google
 ==================================
```
Whoops. Some "=" missing now.

djbrooke replied:
Thanks, corrected in https://github.com/IQSS/dataverse/pull/6192/commits/4931d6afaf2d58d4bca0b4255655c9d43076a882

poikilotherm commented on doc/sphinx-guides/source/installation/oauth2.rst:
```diff
@@ -1,4 +1,4 @@
-OAuth Login: ORCID, GitHub, Google
+OAuth Login: ORCID, Microsoft Azure Active Directory, GitHub, Google
```
Sorry for the noise, but when I see this now, it comes to my mind that this title is going to be included in the TOC and sidebar of the guide. It will be a very long title, getting wrapped. When more providers are added, that list will grow even longer. Is it time to cut off and just name it "OAuth2-based Login" or similar? (Or is there an option to create a short title for TOCs?)

pdurbin replied:
@poikilotherm for a long time I've thought of creating an issue at https://github.com/IQSS/dataverse-ansible/issues called "build guides" or something but I also noticed that you created #6188 and I wonder if that would make things easier. What I'm thinking is that our spin up scripts are now always being executed (thanks to all the effort by @donsizemore 🎉 ) and perhaps we can teach these scripts (by way of dataverse-ansible) to build the guides too so they can be reviewed along with pull requests. Some pull requests **only** touch the guides. 😄

djbrooke replied:
I shortened this, more scalable. https://github.com/IQSS/dataverse/pull/6192/commits/4931d6afaf2d58d4bca0b4255655c9d43076a882

Gerafp commented on .gitignore:
```diff
@@ -60,3 +60,8 @@ scripts/installer/default.config
+#Ignore
+scripts/installer/default.config
+/doc/Architecture/plantuml.jar
+/doc/Architecture/Authentication-oauth.png
```
This modification is personal and I dont ignore when I update the branch. 
The next commit update and delete these lines.

djbrooke commented:
@gerafp and @alejandratenorio - let us know when you've implemented the changes that you plan to do and we'll review this again. Thank you again!

djbrooke commented:
I have one or two more things to check here. Hoping to move over before standup. :)

djbrooke commented:
Thanks all, sorry for the delay. I made one last change to generalize an error message in https://github.com/IQSS/dataverse/pull/6192/commits/ab8714e2f9fbdfb0896ff61907ba83b4702e419e. I'm also fine with the buttons as they are. I'm going to move this over to QA now. I'll take a look at the docs in parallel. 

@kcondon, I tested this successfully with an outlook.com account I created and I also tested this with my Harvard institutional account and the flow worked as I expected, in line with our other Oauth login options. I know that we talked about this a little bit this morning, but if you want to discuss in more detail let me know.

kcondon commented:
Hi @Gerafp , I'm testing this pr and running into a bit of an issue. You've tested both public (microsoft live?) and institutional azure ad/oauth accounts, correct? When you did so, how did you get your client id and secret to configure Dataverse? My understanding from the Azure AD link provided in the docs is one needs to register the app (dataverse), with the Azure AD tenant (directory authenticator/IdP), and that is what gives you the id and secret. If that is the case, I understand about the institutional AD registration and am pursuing that with my local university. However, how does that work with a public service like Live? I'm reading about the /tenant versus the /common userEndpoint but not sure how/where to register to get the right credentials. Thanks for any advice. Kevin P.S. Is this implementation multi-tenant by default? I was looking at this: https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-convert-app-to-be-multi-tenant

Gerafp commented:
> Hi @Gerafp , I'm testing this pr and running into a bit of an issue. You've tested both public (microsoft live?) and institutional azure ad/oauth accounts, correct? When you did so, how did you get your client id and secret to configure Dataverse? My understanding from the Azure AD link provided in the docs is one needs to register the app (dataverse), with the Azure AD tenant (directory authenticator/IdP), and that is what gives you the id and secret. If that is the case, I understand about the institutional AD registration and am pursuing that with my local university. However, how does that work with a public service like Live? I'm reading about the /tenant versus the /common userEndpoint but not sure how/where to register to get the right credentials. Thanks for any advice. Kevin P.S. Is this implementation multi-tenant by default? I was looking at this: https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-convert-app-to-be-multi-tenant

Hi kevin.

You can registry your app in this link 

- https://portal.azure.com/#blade/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/RegisteredApps

Documentation

- https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app

When you registry you app, You obtain the `client_id` and `secret` for testing Dataverse.

djbrooke commented:
Thanks @Gerafp we'll take another look at this on Monday!

kcondon commented:
@Gerafp @djbrooke 

So, I followed the above link and it takes me to the azure portal. My account does not have permission to register an application. This is what I had encountered before.

I had looked at the developer quick start guide earlier and that was helpful in better understanding how the accounts worked and how multi tenant (common) and single tenant (tenantid) endpoints worked to support a variety of microsoft identities.

What I think I learned is this:
1. To configure and use Dataverse with Microsoft Azure AD, you need an Azure subscription that allows creating and managing an Azure AD.
2. You need an Azure AD tenant, either create one or use an existing one
3. You need an Azure account that has permission to register an account for at least one tenant.
4. Once you register Dataverse, you will get a client id and secret. Update your microsoft.json file with  this information.
5. All tenants support both the tenantId and common endpoints, so I think this means you can choose the scope of authentication, whether all Microsoft accounts, regardless of AD (multitenant), or microsoft accounts managed by a single organization (single tenant), by specifying the userEndpoint in microsoft.json, either https://login.microsoftonline.com/{tenantId} or https://login.microsoftonline.com/common. I don't know if login.microsoftonline.com is the literal value or whether it is the tenant name: harvard.microsoftonline.com
6. Now that the microsoft.json file is completed with credentials and endpoint, use the Dataverse api to add Microsoft Azure AD Oauth as an authentication provider.

Are these steps correct? If so, I think we should make the instructions more like these.

djbrooke commented:
I've reached out to our contacts at Harvard to try and set up some test institutional creds.

alejandratenorio commented:
> @Gerafp @djbrooke
> 
> So, I followed the above link and it takes me to the azure portal. My account does not have permission to register an application. This is what I had encountered before.
> 
> I had looked at the developer quick start guide earlier and that was helpful in better understanding how the accounts worked and how multi tenant (common) and single tenant (tenantid) endpoints worked to support a variety of microsoft identities.
> 
> What I think I learned is this:
> 
> 1. To configure and use Dataverse with Microsoft Azure AD, you need an Azure subscription that allows creating and managing an Azure AD.
> 2. You need an Azure AD tenant, either create one or use an existing one. A tenant represents an organization or institution.
> 3. You need an Azure account that has permission to register an application for at least one tenant.
> 4. Once you register Dataverse, you will get a client id and secret. Update your microsoft.json file with  this information.
> 5. All tenants support both the tenantId and common endpoints, so I think this means you can choose the scope of authentication, whether all Microsoft accounts, regardless of AD (multitenant), or microsoft accounts managed by a single organization (single tenant), by specifying the userEndpoint in microsoft.json, either [https://login.microsoftonline.com/{tenantId}](https://login.microsoftonline.com/%7BtenantId%7D) or https://login.microsoftonline.com/common. I don't know if login.microsoftonline.com is the literal value or whether it is the tenant name: harvard.microsoftonline.com
> 6. Now that the microsoft.json file is completed with credentials and endpoint, use the Dataverse api to add Microsoft Azure AD Oauth as an authentication provider.
> 
> Would you confirm these steps are correct? Also, please send the microsoft.json file(s) you used for institutional and public account testing, client id and secret removed.
> 
> If the above directions are correct, I think we should make the instructions more like these.

Hi @kcondon,

What kind of account did you use?  I used my personal account (live.com.mx) and my cgiar account, in both I have permission to register an application.

kcondon commented:
@alejandratenorio When I clicked on the azure portal I was automatically logged in and under switch directory I saw Harvard University as a tenant. Harvard has a subscription, as I understand it, because they provide access to Microsoft cloud applications and we use outlook/web for email. My guess is I have a regular user account as defined by our organization. That is why I was emphasizing the permissions part. I do have a support request in to our IT group for more information. Also, at @pdurbin  suggestion, I created a new microsoft account, accessed the portal and had even less permission -no tenants, no ability to click on switch directories.

Gerafp commented:
> @alejandratenorio When I clicked on the azure portal I was automatically logged in and under switch directory I saw Harvard University as a tenant. Harvard has a subscription, as I understand it, because they provide access to Microsoft cloud applications and we use outlook/web for email. My guess is I have a regular user account as defined by our organization. That is why I was emphasizing the permissions part. I do have a support request in to our IT group for more information. Also, at @pdurbin suggestion, I created a new microsoft account, accessed the portal and had even less permission -no tenants, no ability to click on switch directories.


Hi!

I dont know the reason becausse this happen, Can you give me access to a institutional account for check this situation?.

Personally, I have a institutional account with domain `alumno.buap.mx` and can register a new application in Azure AD.

![image](https://user-images.githubusercontent.com/25359497/67711757-403b3000-f988-11e9-9cde-709a9bd47588.png)

![image](https://user-images.githubusercontent.com/25359497/67711770-4d581f00-f988-11e9-9885-2c90fd062767.png)

djbrooke commented:
Thanks @Gerafp. We unfortunately do not have the ability to edit any of this from our side, as the institutional account here is centrally managed. We're waiting to hear back from support and I'll let you know as soon as I have more information.

kcondon commented:
@alejandratenorio @Gerafp OK, I was able to get the live account to work, thanks for the link. I was also able to get the account I create from Friday working. Not sure why it did not work when I tried it. I can also authenticate my institutional account when I specify the user endpoint: https://login.microsoftonline.com/common but I am unable to register an app with that account -presumably due to a local IT policy. So, this can be merged. Apologies for the confusion. Kevin
